### ISSUE TITLE
Traversal of datapipe graphs broken for pipelines containing SampleMultiplexerDataPipe

### ISSUE THREAD

**Issue by S-t-e-p-h-a-n** [pytorch/pytorch#83975]
## Issue description

Sharding as well as shuffling an iterable datapipe does not work if the datapipe contains a SampleMultiplexerDataPipe at any point after the ShardingFilter.
The root cause for this is that the graph traversal (torch.utils.data.graph.traverse) called by the DataLoader during apply_sharding() cannot traverse past a SampleMultiplexerDataPipe. The reason for this is that during graph traversal getstate_hook() tests whether any of the elements of the object state is an instance of IterDataPipe, MapDataPipe or tuple. In the case of SampleMultiplexerDataPipe the connected pipelines are, however, stored in a dictionary. As a consequence the graph traversal stops at this point and apply_sharding() and apply_shuffling() cannot be called on the relevant elements of the pipeline.

## Code example

    from torch.utils.data import DataLoader
    from torchdata.datapipes.iter import IterableWrapper, SampleMultiplexer

    pipe1 = IterableWrapper(range(2))
    pipe1 = pipe1.sharding_filter()
    pipe1 = pipe1.shuffle()

    pipe2 = IterableWrapper(range(2, 4))
    pipe2 = pipe2.sharding_filter()
    pipe2 = pipe2.shuffle()

    pipe = SampleMultiplexer({pipe1: 0.5, pipe2: 0.5})

    data_loader = DataLoader(pipe, shuffle=True, num_workers=2)

    for sample in data_loader:
        print(sample)

Running this code produces the following warning, even though the pipelines contain two Shufflers:

    UserWarning: `shuffle=True` was set, but the datapipe does not contain a `Shuffler`. Adding one at the end. Be aware that the default buffer size might not be sufficient for your task.

This indicates that Shuffle was not found during graph traversal.
Furthermore, each element produced by the dataset is emitted n times, where n = num_workers. This indicates that sharding is not working.

Expected output:
- no warning about missing Shuffle
- tensors with the values 0, 1, 2, and 3 in an arbitrary order with each number printed exactly once

## System Info
PyTorch version: 1.12.1+cu116
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Gentoo Base System release 2.8 (x86_64)
GCC version: (Gentoo 11.3.0 p4) 11.3.0
Clang version: 14.0.6
CMake version: version 3.22.4
Libc version: glibc-2.35

Python version: 3.10.5 (main, Jun 23 2022, 09:37:51) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.59-gentoo-x86_64-Intel-R-_Xeon-R-_E-2146G_CPU_@_3.50GHz-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 11.6.124
GPU models and configuration: 
GPU 0: NVIDIA GeForce GTX 1050 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/opt/cuda/targets/x86_64-linux/lib/libcudnn.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_adv_infer.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_adv_train.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_cnn_infer.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_cnn_train.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_ops_infer.so.8.4.1
/opt/cuda/targets/x86_64-linux/lib/libcudnn_ops_train.so.8.4.1
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.1
[pip3] torch==1.12.1+cu116
[pip3] torchdata==0.4.1
[pip3] torchvision==0.13.1+cu116
[conda] Could not collect

Process finished with exit code 0
